src/deploy/functions/cache/hash.ts

```diff
@@ -0,0 +1,55 @@
+import * as fs from "fs";
+import * as crypto from "crypto";
+import * as secrets from "../../../functions/secrets";
+import { Backend, Endpoint, EnvironmentVariables, allEndpoints } from "../backend";
+
+/**
+ * Retrieves the unique hash given a {@link Backend} and sourceFile.
+ * @param backend Backend of a set of functions
+ * @param sourceFile Packaged file contents of functions
+ */
+export async function getBackendHash(backend: Backend, sourceFile?: string) {
+  const hash = crypto.createHash("sha256");
+
+  // If present, hash the contents of the source file
+  if (sourceFile) {
+    await new Promise((resolve, reject) => {
+      const readStream = fs.createReadStream(sourceFile);
+      readStream.on("error", (err) => reject(err));
+      readStream.on("data", (chunk) => chunk && hash.push(chunk));
+      readStream.on("end", () => resolve(hash.digest("hex")));
```

Comment:
nit* I don't think this digest is necessary.

Reply:
Well, it's ambiguous since the Promise isn't annotated as returning anything and IDK if TypeScript can deduce it here. I personally think that this Promise should return the hash, not the digest, so that the env vars and secrets configs can be pushed onto the same Hash object. I think I would probably create a utility for adding a file to a hash and then call something like:

```
const hash = crypto.createHash("sha256");
await addFileTohash(sourceFile, hash);
await addFileToHash(.env, hash);
await addFileTohash(.env.alias, hash)
await addFileToHash(.env.alias, hash)
hash.push(JSON.stringify(endpoint.secrets));
```

Though that seems super wasteful becuase the source & dotenv hashes are really expensive to generate (file I/O) and are the same for an entire backend. So I'm starting to think that you can do a concatination of multiple hashes as long as we can guarantee that the string doesn't get too long (and if it does... I guess we can hash it again?).

So for a whole project you'll share a hash of the env files. And within a codbase, all v1 functions share one source hash and all v2 functions share another source hash. Then finally, each endpoint will have a hash based on the config of the secrets it depends upon.

But hashes are themselves pretty long and we may not be able to hold a whole hash in a label value, so... maybe an endpoint's hash is `hash(hash(env) + hash(source) + hash(secrets))`

Reply:
Updated the generation of the hashes to be `hash(hash(env) + hash(source) + hash(secrets))`.

#4876 will be updated to reflect this shortly